planner/cascades/testdata/transformation_rules_suite_out.json

```diff
@@ -2456,5 +2456,212 @@
+          "Group#1 Schema:[test.t.a,test.t.b,test.t.c,test.t.d,Column#14,Column#17,Column#18]",
+          "    Window_10 input:[Group#2]",
+          "Group#2 Schema:[test.t.a,test.t.b,test.t.c,test.t.d,Column#14,Column#17]",
+          "    Window_12 input:[Group#3]",
+          "Group#3 Schema:[test.t.a,test.t.b,test.t.c,test.t.d]",
+          "    DataSource_1 table:t"
+        ]
+      },
+      {
+        "SQL": "select a, b, sum(bb) over (partition by a) as 'sum_bb' from (select a, b, max(b) over (partition by a) as 'bb' from t) as tt",
+        "Result": [
+          "Group#0 Schema:[test.t.a,test.t.b,Column#19]",
+          "    Projection_11 input:[Group#1], test.t.a, test.t.b, Column#16",
+          "Group#1 Schema:[test.t.a,test.t.b,Column#14,Column#16]",
+          "    Window_8 input:[Group#2]",
+          "Group#2 Schema:[test.t.a,test.t.b,Column#14]",
+          "    Window_4 input:[Group#3]",
+          "Group#3 Schema:[test.t.a,test.t.b]",
+          "    DataSource_1 table:t"
+        ]
+      },
+      {
+        "SQL": "select a, b, sum(bb) over (partition by a) as 'sum_bb', c from (select a, b, c, max(b) over (partition by a) as 'bb' from t) as tt",
+        "Result": [
+          "Group#0 Schema:[test.t.a,test.t.b,Column#19,test.t.c]",
+          "    Projection_11 input:[Group#1], test.t.a, test.t.b, Column#16, test.t.c",
+          "Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#16]",
+          "    Window_8 input:[Group#2]",
+          "Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#14]",
+          "    Window_4 input:[Group#3]",
+          "Group#3 Schema:[test.t.a,test.t.b,test.t.c]",
+          "    DataSource_1 table:t"
+        ]
+      },
+      {
+        "SQL": "select a, b, rank() over (partition by a), sum(bb) over (partition by a) as 'sum_bb', c from (select a, b, c, max(b) over (partition by a) as 'bb' from t) as tt",
+        "Result": [
+          "Group#0 Schema:[test.t.a,test.t.b,Column#21,Column#22,test.t.c]",
+          "    Projection_13 input:[Group#1], test.t.a, test.t.b, Column#18, Column#17, test.t.c",
+          "Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#17,Column#18]",
+          "    Window_10 input:[Group#2]",
+          "Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#17]",
+          "    Projection_9 input:[Group#3], test.t.a, test.t.b, test.t.c, Column#17",
+          "Group#3 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17]",
+          "    Window_8 input:[Group#4]",
+          "Group#4 Schema:[test.t.a,test.t.b,test.t.c,Column#14]",
+          "    Window_4 input:[Group#5]",
+          "Group#5 Schema:[test.t.a,test.t.b,test.t.c]",
+          "    DataSource_1 table:t"
+        ]
+      },
+      {
+        "SQL": "select a, b, sum(bb) over (partition by a) as 'sum_bb', c, rank() over (partition by a) from (select a, b, c, max(b) over (partition by a) as 'bb' from t) as tt",
```

Comment:
I think the two windows should not be merged in this test?

Reply:
The SQL:
```
select a, b, rank() over (partition by a), sum(bb) over (partition by a) as 'sum_bb', c from (select a, b, c, max(b) over (partition by a) as 'bb' from t) as tt
```
Before cascades:
```
Group#0 Schema:[test.t.a,test.t.b,Column#21,Column#22,test.t.c]
    Projection_12 input:[Group#1], test.t.a, test.t.b, Column#18, Column#17, test.t.c
Group#1 Schema:[test.t.a,test.t.b,Column#18,Column#17,test.t.c]
    Projection_11 input:[Group#2], test.t.a, test.t.b, Column#18, Column#17, test.t.c
Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#17,Column#18]
    Window_10 input:[Group#3]
Group#3 Schema:[test.t.a,test.t.b,test.t.c,Column#17]
    Projection_9 input:[Group#4], test.t.a, test.t.b, test.t.c, Column#17
Group#4 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17]
    Window_8 input:[Group#5]
Group#5 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_7 input:[Group#6], test.t.a, test.t.b, test.t.c, Column#14
Group#6 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_6 input:[Group#7], test.t.a, test.t.b, test.t.c, Column#14
Group#7 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Projection_5 input:[Group#8], test.t.a, test.t.b, test.t.c, Column#14
Group#8 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
    Window_4 input:[Group#9]
Group#9 Schema:[test.t.a,test.t.b,test.t.c]
    Projection_3 input:[Group#10], test.t.a, test.t.b, test.t.c
Group#10 Schema:[test.t.a,test.t.b,test.t.c]
    Projection_2 input:[Group#11], test.t.a, test.t.b, test.t.c
Group#11 Schema:[test.t.a,test.t.b,test.t.c]
    DataSource_1 table:t
```

After cascades (rules `EliminateProjection` and `MergeAdjacentProjection`):
```
Group#0 Schema:[test.t.a,test.t.b,Column#21,Column#22,test.t.c]
     Projection_13 input:[Group#1], test.t.a, test.t.b, Column#18, Column#17, test.t.c 
Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#17,Column#18]
     Window_10 input:[Group#2] 
Group#2 Schema:[test.t.a,test.t.b,test.t.c,Column#17]
     Projection_9 input:[Group#3], test.t.a, test.t.b, test.t.c, Column#17 
Group#3 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17]
     Window_8 input:[Group#4] 
Group#4 Schema:[test.t.a,test.t.b,test.t.c,Column#14]
     Window_4 input:[Group#5] 
Group#5 Schema:[test.t.a,test.t.b,test.t.c]
     DataSource_1 table:t]
```

I think the Group#2 Projection should be eliminated by  `EliminateProjection` or other rules. 

`MergeAdjacentWindow` can only merge adjacent window groups so that the two window function cannot be merged.  Isn't that true?

Reply:
Emmm, you may misunderstand me. 

The SQL:
```
select a, b, sum(bb) over (partition by a) as 'sum_bb', c, rank() over (partition by a) 
from (
    select a, b, c, max(b) over (partition by a) as 'bb' 
    from t
) as tt
```

I think this SQL should generate a plan with two Windows.  Because the outer one reference `bb`, which is the result of the inner one. So the two windows can not be merged.

But the test result only contains one:
```


"Result": [
          "Group#0 Schema:[test.t.a,test.t.b,Column#21,test.t.c,Column#23]",
          "    Projection_15 input:[Group#1], test.t.a, test.t.b, Column#18, test.t.c, Column#17",
          "Group#1 Schema:[test.t.a,test.t.b,test.t.c,Column#14,Column#17,Column#18]",
          "    Window_14 input:[Group#2]",          "Group#2 Schema:[test.t.a,test.t.b,test.t.c]",
          "    DataSource_1 table:t"
```

Reply:
Thanks for your guidance! This bug is produced by the abuse of  `LogicalPlan` schema. 😸  I fix it in new commit just now.